MlDistributedFailureIT#testJobRelocationIsMemoryAware failed

**Build scan**: https://gradle-enterprise.elastic.co/s/zwqvn6whenbjm

**Repro line**: `./gradlew ':x-pack:plugin:ml:internalClusterTest' --tests "org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT.testJobRelocationIsMemoryAware" -Dtests.seed=59B110C189FF078D -Dtests.security.manager=true -Dtests.locale=be-BY -Dtests.timezone=AET -Druntime.java=11`

**Reproduces locally?**: Tried a few times, did not see any failures.

**Applicable branches**:

[**Failure history**](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-90d,mode:quick,to:now))&_a=(columns:!(build.branch),index:b646ed00-7efc-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:testJobRelocationIsMemoryAware),sort:!(process.time-start,desc))): So far just once on `master` and another on a PR build.

**Failure excerpt**:

```
org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT > testJobRelocationIsMemoryAware FAILED
    java.lang.AssertionError: expected:<1> but was:<2>
        at __randomizedtesting.SeedInfo.seed([59B110C189FF078D:CAF4F9484DD41423]:0)
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT.lambda$testJobRelocationIsMemoryAware$14(MlDistributedFailureIT.java:414)
        at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:942)
        at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:915)
        at org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT.testJobRelocationIsMemoryAware(MlDistributedFailureIT.java:401)
```

The test seems to be horrifically slowed down by this problem occurring several times per second:

```
1> [2020-10-11T15:42:23,325][ERROR][o.e.x.c.t.IndexTemplateRegistry] [node_t0] error adding index template [.slm-history] from [/slm-history.json] for [index_lifecycle] |  
-- | --
  | 1> java.lang.IllegalArgumentException: composable template [.slm-history] template after composition is invalid |  
  | 1> 	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.addIndexTemplateV2(MetadataIndexTemplateService.java:521) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService$4.execute(MetadataIndexTemplateService.java:417) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:47) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:697) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:319) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:214) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:151) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:674) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?] |  
  | 1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?] |  
  | 1> 	at java.lang.Thread.run(Thread.java:834) [?:?] |  
  | 1> Caused by: org.elasticsearch.common.xcontent.XContentParseException: [index_template] unknown field [data_stream] |  
  | 1> 	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.lambda$validateCompositeTemplate$21(MetadataIndexTemplateService.java:1150) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.indices.IndicesService.withTempIndexService(IndicesService.java:621) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.validateCompositeTemplate(MetadataIndexTemplateService.java:1130) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	at org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.addIndexTemplateV2(MetadataIndexTemplateService.java:517) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | 1> 	... 14 more
```

I think this isn't really a problem in ML, but some interaction of SLM and data streams making everything run really slowly.

--- a/x-pack/plugin/ml/build.gradle
+++ b/x-pack/plugin/ml/build.gradle
@@ -50,6 +50,7 @@ dependencies {
   compileOnly project(path: xpackModule('core'), configuration: 'default')
   testImplementation project(path: xpackModule('core'), configuration: 'testArtifacts')
   testImplementation project(path: xpackModule('ilm'), configuration: 'default')
+  testImplementation project(path: xpackModule('data-streams'), configuration: 'default')
   // This should not be here
   testImplementation project(path: xpackModule('security'), configuration: 'testArtifacts')
 

--- a/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/support/BaseMlIntegTestCase.java
+++ b/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/support/BaseMlIntegTestCase.java
@@ -57,6 +57,7 @@ import org.elasticsearch.xpack.core.ml.job.config.Detector;
 import org.elasticsearch.xpack.core.ml.job.config.Job;
 import org.elasticsearch.xpack.core.ml.job.config.JobState;
 import org.elasticsearch.xpack.core.ml.job.process.autodetect.state.DataCounts;
+import org.elasticsearch.xpack.datastreams.DataStreamsPlugin;
 import org.elasticsearch.xpack.ilm.IndexLifecycle;
 import org.elasticsearch.xpack.ml.LocalStateMachineLearning;
 import org.elasticsearch.xpack.ml.MachineLearning;
@@ -112,7 +113,9 @@ public abstract class BaseMlIntegTestCase extends ESIntegTestCase {
             CommonAnalysisPlugin.class,
             ReindexPlugin.class,
             // ILM is required for .ml-state template index settings
-            IndexLifecycle.class);
+            IndexLifecycle.class,
+            // Deprecation warnings go to a data stream, if we ever cause a deprecation warning the data streams plugin is required
+            DataStreamsPlugin.class);
     }
 
     @Override

